### ISSUE TITLE
Misconfiguration of blocking IO dispatcher for akka streams

### ISSUE THREAD

**Issue by nick-nachos** [akka/akka#24357]
While working with akka streams API, I came across an issue when I used the following statement:
```scala
someSink.withAttributes(ActorAttributes.dispatcher("akka.stream.blocking-io-dispatcher"))
```
That call resulted to the following exception:

```
Caused by: com.typesafe.config.ConfigException$WrongType: reference.conf @ jar:file:/home/nnakas/.m2/repository/com/typesafe/akka/akka-stream_2.11/2.5.6/akka-stream_2.11-2.5.6.jar!/reference.conf: 80: akka.stream.blocking-io-dispatcher has type STRING rather than OBJECT
	at com.typesafe.config.impl.SimpleConfig.findKeyOrNull(SimpleConfig.java:159)
	at com.typesafe.config.impl.SimpleConfig.findOrNull(SimpleConfig.java:170)
	at com.typesafe.config.impl.SimpleConfig.findOrNull(SimpleConfig.java:176)
	at com.typesafe.config.impl.SimpleConfig.findOrNull(SimpleConfig.java:176)
	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:184)
	at com.typesafe.config.impl.SimpleConfig.find(SimpleConfig.java:189)
	at com.typesafe.config.impl.SimpleConfig.getObject(SimpleConfig.java:264)
	at com.typesafe.config.impl.SimpleConfig.getConfig(SimpleConfig.java:270)
	at com.typesafe.config.impl.SimpleConfig.getConfig(SimpleConfig.java:37)
	at akka.dispatch.Dispatchers.config(Dispatchers.scala:127)
	at akka.dispatch.Dispatchers.lookupConfigurator(Dispatchers.scala:96)
	at akka.dispatch.Dispatchers.lookup(Dispatchers.scala:79)
	at akka.actor.LocalActorRefProvider.actorOf(ActorRefProvider.scala:746)
	... 67 more
```

I went through the akka sources, and the configuration at reference.conf is as follows:

```
# Fully qualified config path which holds the dispatcher configuration
# to be used by ActorMaterializer when creating Actors for IO operations,
# such as FileSource, FileSink and others.
blocking-io-dispatcher = "akka.stream.default-blocking-io-dispatcher"

default-blocking-io-dispatcher {
  type = "Dispatcher"
  executor = "thread-pool-executor"
  throughput = 1

  thread-pool-executor {
    fixed-pool-size = 16
  }
}
```

So it would seem that the configuration should probably be like this:
```
blocking-io-dispatcher = ${akka.stream.default-blocking-io-dispatcher}
```

This finding however made me wonder **why this error doesn't occur in the akka-stream library in general**. Theoretically, according to the [akka docs](https://doc.akka.io/docs/akka/current/stream/stream-io.html?language=scala#streaming-file-io):
> If you want to configure a custom dispatcher for file IO operations globally, you can do so by changing the akka.stream.blocking-io-dispatcher

so I went a little bit further with the research, and came across this inside `akka.stream.ActorAttributes`:
```scala
object ActorAttributes {
  import Attributes._
  final case class Dispatcher(dispatcher: String) extends Attribute
  final case class SupervisionStrategy(decider: Supervision.Decider) extends Attribute

  val IODispatcher: Dispatcher = ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")
  // ...
```
This effectively means that the stream IO dispatcher is fixed to the default one (i.e. `akka.stream.default-blocking-io-dispatcher`) and cannot be overridden via `akka.stream.blocking-io-dispatcher` setting, which is the reason however that the dispatcher resolution works in the first place (`default-blocking-io-dispatcher` is a valid Config object as demonstrated above)

To sum up, the issues that need to be tackled are as follows:
1. Fix the resolution of the `akka.stream.blocking-io-dispatcher` property
2. Use `akka.stream.blocking-io-dispatcher` instead of `akka.stream.default-blocking-io-dispatcher` within the akka-stream library.

If you approve of the issue and the suggested approach for fixing this, I'll be happy to PR the fix.

**Comment by johanandren** [comment]
Good catch!

The reason why it was not noticed that the `default-blocking-io-dispatcher` roundtrip does not work is that all code use the `akka.stream.ActorAttributes.IODispatcher` constant, which points to `akka.stream.default-blocking-io-dispatcher` (in addition to missing test coverage for the feature).

The correct fix for this would be to leave the config as is, change that `IODispatcher` constant to point to the `akka.stream.blocking-io-dispatcher` and add translation logic that matches against that and picks up the right dispatcher name from config somewhere in the materializer.
